[AArch64][SME] Fix frame lowering not using a base pointer for SME functions.

aemerson opened the pull request:
The existing code is checking for the presence of the +sve subtarget feature
when deciding to use a base pointer for the function, but this check doesn't
work when only +sme is used.

rdar://126878490

efriedma-quic commented:
Please add a testcase where a base pointer actually improves the generated code.  None of the testcases you're modifying even materialize the base pointer.

(I think some of the existing testcases are likely to be changed to add +sve; see discussion on #83301 .)

aemerson commented:
> Please add a testcase where a base pointer actually improves the generated code. None of the testcases you're modifying even materialize the base pointer.
> 
> (I think some of the existing testcases are likely to be changed to add +sve; see discussion on #83301 .)

The test case I have is a pretty large one that causes a crash. I'll add it but that was the only motivation for this, not to improve code.

aemerson committed:
[AArch64][SME] Fix frame lowering not using a base pointer for SME functions.

The existing code is checking for the presence of the +sve subtarget feature
when deciding to use a base pointer for the function, but this check doesn't
work when only +sme is used.

rdar://126878490

efriedma-quic commented:
Your testcase doesn't appear to crash on trunk?  Or is that reduced?

If you just want to force the codepath in question, something smaller like the following works:

```
void g(int, void*, svfloat32_t, void*);
void f(int x, svfloat32_t y) {
  void *p = __builtin_alloca(x);
  __attribute((aligned(16))) int a[32]; 
  asm("":::"v0","v1","v2","v3","v4","v5","v6","v7","v8","v9","v10","v11","v12","v13","v14","v15","v16",
    "v17","v18","v19","v20","v21","v22","v23","v24","v25","v26","v27","v28","v29","v30","v31",
    "x0","x1","x2","x3","x4","x5","x6","x7","x8","x9","x10","x11","x12","x13","x14","x15","x16",
    "x17","x18","x19","x20","x21","x22","x23","x24","x25","x26","x27","x28","x30");
  g(x,p,y,a);
}
```

----

Note we'll probably end up messing with the fp/bp code at some point to address #80009.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -552,7 +552,8 @@ bool AArch64RegisterInfo::hasBasePointer(const MachineFunction &MF) const {
 
-    if (MF.getSubtarget<AArch64Subtarget>().hasSVE()) {
+    auto &ST = MF.getSubtarget<AArch64Subtarget>();
+    if (ST.hasSVEorSME()) {
```
Really what you want is `ST.hasSVE() || (ST.hasSME() && SMEAttrs(MF->getFunction).hasStreamingInterfaceOrBody())`. Maybe the subtarget should expose a helper for this.

aemerson commented:
> Your testcase doesn't appear to crash on trunk? Or is that reduced?
> 
> If you just want to force the codepath in question, something smaller like the following works:
> 
> ```
> void g(int, void*, svfloat32_t, void*);
> void f(int x, svfloat32_t y) {
>   void *p = __builtin_alloca(x);
>   __attribute((aligned(16))) int a[32]; 
>   asm("":::"v0","v1","v2","v3","v4","v5","v6","v7","v8","v9","v10","v11","v12","v13","v14","v15","v16",
>     "v17","v18","v19","v20","v21","v22","v23","v24","v25","v26","v27","v28","v29","v30","v31",
>     "x0","x1","x2","x3","x4","x5","x6","x7","x8","x9","x10","x11","x12","x13","x14","x15","x16",
>     "x17","x18","x19","x20","x21","x22","x23","x24","x25","x26","x27","x28","x30");
>   g(x,p,y,a);
> }
> ```
> 
> Note we'll probably end up messing with the fp/bp code at some point to address #80009.

Ah I forgot to add -O0 to the RUN line.

efriedma-quic commented:
Oh, I see, reproduced.

Probably there should be a comment in this code that the base pointer is necessary to ensure the emergency spill slot is reachable.

Probably you can write a reduced testcase to force an emergency spill (just need to ensure you have a store to a stack slot with no scavengeable register... should be able to force that with inline asm).  I'd be okay leaving it out if it turns out to be too tricky to write, though.  The unreduced case probably isn't that useful.

aemerson committed:
Add MIR test.

aemerson commented:
> Oh, I see, reproduced.
> 
> Probably there should be a comment in this code that the base pointer is necessary to ensure the emergency spill slot is reachable.
> 
> Probably you can write a reduced testcase to force an emergency spill (just need to ensure you have a store to a stack slot with no scavengeable register... should be able to force that with inline asm). I'd be okay leaving it out if it turns out to be too tricky to write, though. The unreduced case probably isn't that useful.

I wasn't able to write a smaller test for the crash (probably did something wrong) but there was an existing SVE test for BP that I've replicated for SME.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -552,7 +552,9 @@ bool AArch64RegisterInfo::hasBasePointer(const MachineFunction &MF) const {
-    if (MF.getSubtarget<AArch64Subtarget>().hasSVE()) {
+    auto &ST = MF.getSubtarget<AArch64Subtarget>();
+    if (ST.hasSVE() ||
+        (ST.hasSME() && ST.hasStreamingInterfaceOrBody(MF.getFunction()))) {
```
Oops, didn't realize there was an existing bit for this... `ST,isStreaming()` is equivalent, I think.

aemerson replied:
Don't we need to check it on a per Function basis?

efriedma-quic replied:
The subtarget is per-function, sort of... see AArch64TargetMachine::getSubtargetImpl.

aemerson replied:
Oh right, of course.

aemerson committed:
Fixups

efriedma-quic commented on llvm/test/CodeGen/AArch64/framelayout-sme-basepointer.mir:
```diff
@@ -0,0 +1,28 @@
+...
+---
+# This test verifies that the basepointer is available in presence of SME stack objects.
+# This is almost identical to the SVE version of this test but has to be a separate file
```
Can you just mark the target attribute on the IR function, instead of using -mattr?  Not a big deal either way.

efriedma-quic reviewed (approved):
LGTM

aemerson committed:
Merge tests.

aemerson committed:
Fix CHECK prefixes.
